File: modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-02-preparacao-datasets/dataset-preparation-pipeline-java/src/test/java/com/datasetprep/cleaning/MinHashLshTest.java
package com.datasetprep.cleaning;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class MinHashLshTest {

    private Exemplo achar(List<Exemplo> dataset, String id) {
        return dataset.stream().filter(e -> e.metadata().id().equals(id)).findFirst().orElseThrow();
    }

    @Test
    void assinaturaMinHashDeTextoIdenticoProduzSimilaridadeEstimada1() {
        List<Exemplo> dataset = SampleDataset.gerarDatasetSimulado();
        var coeficientes = MinHashLsh.gerarCoeficientesHash(MinHashLsh.MINHASH_K, MinHashLsh.MINHASH_SEMENTE);
        Exemplo e = achar(dataset, "amplitude-auto-Oficina Estrela-0");
        long[] sig = MinHashLsh.assinaturaMinHash(MinHashLsh.shingles(e.entrada(), 5), coeficientes);
        assertThat(MinHashLsh.similaridadeMinHashEstimada(sig, sig)).isEqualTo(1.0);
    }

    @Test
    void minHashAproximaJaccardExatoParaDuplicataExata() {
        List<Exemplo> dataset = SampleDataset.gerarDatasetSimulado();
        var coeficientes = MinHashLsh.gerarCoeficientesHash(MinHashLsh.MINHASH_K, MinHashLsh.MINHASH_SEMENTE);
        Exemplo a = achar(dataset, "amplitude-auto-Oficina Estrela-0");
        Exemplo b = achar(dataset, "amplitude-auto-Oficina Estrela-0-reenviado");
        long[] sigA = MinHashLsh.assinaturaMinHash(MinHashLsh.shingles(a.entrada(), 5), coeficientes);
        long[] sigB = MinHashLsh.assinaturaMinHash(MinHashLsh.shingles(b.entrada(), 5), coeficientes);
        double estimado = MinHashLsh.similaridadeMinHashEstimada(sigA, sigB);
        double exato = MinHashLsh.similaridadeJaccardExata(a.entrada(), b.entrada());
        assertThat(Math.abs(estimado - exato)).isLessThan(0.1);
    }

    @Test
    void minHashAproximaJaccardExatoParaDuplicataComRuidoDeOcr() {
        List<Exemplo> dataset = SampleDataset.gerarDatasetSimulado();
        var coeficientes = MinHashLsh.gerarCoeficientesHash(MinHashLsh.MINHASH_K, MinHashLsh.MINHASH_SEMENTE);
        Exemplo a = achar(dataset, "amplitude-auto-Oficina Estrela-2");
        Exemplo b = achar(dataset, "amplitude-auto-Oficina Estrela-2-ruido-ocr");
        long[] sigA = MinHashLsh.assinaturaMinHash(MinHashLsh.shingles(a.entrada(), 5), coeficientes);
        long[] sigB = MinHashLsh.assinaturaMinHash(MinHashLsh.shingles(b.entrada(), 5), coeficientes);
        double estimado = MinHashLsh.similaridadeMinHashEstimada(sigA, sigB);
        double exato = MinHashLsh.similaridadeJaccardExata(a.entrada(), b.entrada());
        assertThat(Math.abs(estimado - exato)).isLessThan(0.15);
    }

    @Test
    void lshEncontraExatamenteOs3ParesDeQuaseDuplicataPlantados() {
        List<Exemplo> dataset = SampleDataset.gerarDatasetSimulado();
        var r = MinHashLsh.encontrarQuaseDuplicatasMinHashLSH(dataset);
        assertThat(r.paresDuplicata()).hasSize(3);
    }

    @Test
    void lshReduzOsNumeroDeComparacoesEmPeloMenos80Porcento() {
        List<Exemplo> dataset = SampleDataset.gerarDatasetSimulado();
        var r = MinHashLsh.encontrarQuaseDuplicatasMinHashLSH(dataset);
        double reducao = 1 - (double) r.totalCandidatosLSH() / r.totalParesForcaBruta();
        assertThat(reducao).isGreaterThanOrEqualTo(0.8);
    }

    @Test
    void nenhumParNaoDuplicataEConfirmadoComoDuplicata() {
        List<Exemplo> dataset = SampleDataset.gerarDatasetSimulado();
        var r = MinHashLsh.encontrarQuaseDuplicatasMinHashLSH(dataset);
        List<String> idsDuplicata = new ArrayList<>();
        for (var par : r.paresDuplicata()) {
            idsDuplicata.add(dataset.get(par.i()).metadata().id());
            idsDuplicata.add(dataset.get(par.j()).metadata().id());
        }
        assertThat(idsDuplicata).containsExactlyInAnyOrder(
                "amplitude-auto-Oficina Estrela-0", "amplitude-auto-Oficina Estrela-0-reenviado",
                "amplitude-auto-Oficina Estrela-2", "amplitude-auto-Oficina Estrela-2-ruido-ocr",
                "amplitude-saude-empresarial-Clinica Vitalis-0", "amplitude-saude-empresarial-Clinica Vitalis-0-reenviado");
    }

    @Test
    void remocaoDeQuaseDuplicataTiraExatamente3Exemplos() {
        List<Exemplo> dataset = SampleDataset.gerarDatasetSimulado();
        var r = MinHashLsh.removerQuaseDuplicatas(dataset);
        assertThat(r.removidos()).isEqualTo(3);
    }
}

File: modulo09-processamento-de-dados-e-fine-tuning-de-modelos/modulo-02-preparacao-datasets/dataset-preparation-pipeline-go/cleaning/minhashlsh_test.go
package cleaning

import (
	"math"
	"testing"
)

func acharExemplo(dataset []Exemplo, id string) Exemplo {
	for _, e := range dataset {
		if e.Metadata.ID == id {
			return e
		}
	}
	panic("nao encontrado: " + id)
}

func TestAssinaturaMinHashDeTextoIdenticoProduzSimilaridade1(t *testing.T) {
	dataset := GerarDatasetSimulado()
	coeficientes := GerarCoeficientesHash(MinhashK, MinhashSemente)
	e := acharExemplo(dataset, "amplitude-auto-Oficina Estrela-0")
	sig := AssinaturaMinHash(Shingles(e.Entrada, 5), coeficientes)
	if SimilaridadeMinHashEstimada(sig, sig) != 1.0 {
		t.Fatal("esperado similaridade 1.0 contra si mesmo")
	}
}

func TestMinHashAproximaJaccardExatoParaDuplicataExata(t *testing.T) {
	dataset := GerarDatasetSimulado()
	coeficientes := GerarCoeficientesHash(MinhashK, MinhashSemente)
	a := acharExemplo(dataset, "amplitude-auto-Oficina Estrela-0")
	b := acharExemplo(dataset, "amplitude-auto-Oficina Estrela-0-reenviado")
	sigA := AssinaturaMinHash(Shingles(a.Entrada, 5), coeficientes)
	sigB := AssinaturaMinHash(Shingles(b.Entrada, 5), coeficientes)
	estimado := SimilaridadeMinHashEstimada(sigA, sigB)
	exato := SimilaridadeJaccardExata(a.Entrada, b.Entrada, 5)
	if math.Abs(estimado-exato) >= 0.1 {
		t.Fatalf("erro muito alto: exato=%v estimado=%v", exato, estimado)
	}
}

func TestMinHashAproximaJaccardExatoParaDuplicataComRuidoDeOcr(t *testing.T) {
	dataset := GerarDatasetSimulado()
	coeficientes := GerarCoeficientesHash(MinhashK, MinhashSemente)
	a := acharExemplo(dataset, "amplitude-auto-Oficina Estrela-2")
	b := acharExemplo(dataset, "amplitude-auto-Oficina Estrela-2-ruido-ocr")
	sigA := AssinaturaMinHash(Shingles(a.Entrada, 5), coeficientes)
	sigB := AssinaturaMinHash(Shingles(b.Entrada, 5), coeficientes)
	estimado := SimilaridadeMinHashEstimada(sigA, sigB)
	exato := SimilaridadeJaccardExata(a.Entrada, b.Entrada, 5)
	if math.Abs(estimado-exato) >= 0.15 {
		t.Fatalf("erro muito alto: exato=%v estimado=%v", exato, estimado)
	}
}

func TestLshEncontraExatamenteOs3ParesDeQuaseDuplicataPlantados(t *testing.T) {
	dataset := GerarDatasetSimulado()
	r := EncontrarQuaseDuplicatasMinHashLSH(dataset)
	if len(r.ParesDuplicata) != 3 {
		t.Fatalf("esperado 3 pares, obtive %d", len(r.ParesDuplicata))
	}
}

func TestLshReduzComparacoesEmPeloMenos80Porcento(t *testing.T) {
	dataset := GerarDatasetSimulado()
	r := EncontrarQuaseDuplicatasMinHashLSH(dataset)
	reducao := 1 - float64(r.TotalCandidatosLSH)/float64(r.TotalParesForcaBruta)
	if reducao < 0.8 {
		t.Fatalf("reducao de so %.1f%%", reducao*100)
	}
}

func TestNenhumParNaoDuplicataEConfirmado(t *testing.T) {
	dataset := GerarDatasetSimulado()
	r := EncontrarQuaseDuplicatasMinHashLSH(dataset)
	idsDuplicata := map[string]bool{}
	for _, p := range r.ParesDuplicata {
		idsDuplicata[dataset[p.I].Metadata.ID] = true
		idsDuplicata[dataset[p.J].Metadata.ID] = true
	}
	esperados := []string{
		"amplitude-auto-Oficina Estrela-0", "amplitude-auto-Oficina Estrela-0-reenviado",
		"amplitude-auto-Oficina Estrela-2", "amplitude-auto-Oficina Estrela-2-ruido-ocr",
		"amplitude-saude-empresarial-Clinica Vitalis-0", "amplitude-saude-empresarial-Clinica Vitalis-0-reenviado",
	}
	if len(idsDuplicata) != len(esperados) {
		t.Fatalf("esperado %d ids, obtive %d: %v", len(esperados), len(idsDuplicata), idsDuplicata)
	}
	for _, id := range esperados {
		if !idsDuplicata[id] {
			t.Errorf("id esperado ausente: %s", id)
		}
	}
}

func TestRemocaoDeQuaseDuplicataTiraExatamente3(t *testing.T) {
	dataset := GerarDatasetSimulado()
	r := RemoverQuaseDuplicatas(dataset)
	if r.Removidos != 3 {
		t.Fatalf("esperado 3 removidos, obtive %d", r.Removidos)
	}
}
